File: src/python/core/infrastructure/llm/extraction/llm_output_extractor.py
"""Pure utility for cleaning and extracting JSON from raw LLM output."""

import re

NO_JSON_FOUND_MESSAGE = "No JSON object found in LLM output"


def extract_json(raw_output: str) -> str:
    """
    Cleans raw LLM output and extracts the outermost JSON object as a string.
    
    Handles:
    - Markdown code fences (```json ... ```)
    - <think>...</think> blocks (DeepSeek-R1)
    - Preamble/postamble text
    
    Args:
        raw_output: Raw text from LLM
        
    Returns:
        Extracted JSON string
        
    Raises:
        ValueError: If no valid JSON object found
    """
    if not raw_output or not raw_output.strip():
        raise ValueError(NO_JSON_FOUND_MESSAGE)
    
    cleaned = strip_thinking_blocks(raw_output)
    cleaned = strip_markdown_fences(cleaned)
    
    return extract_outermost_json_object(cleaned)


def strip_thinking_blocks(text: str) -> str:
    """Remove <think>...</think> blocks (e.g. DeepSeek-R1 reasoning)."""
    return re.sub(r'<think>[\s\S]*?</think>', '', text, flags=re.IGNORECASE).strip()


def strip_markdown_fences(text: str) -> str:
    """Remove markdown code fences: ```json ... ``` or ``` ... ```."""
    return re.sub(r'```(?:json)?\s*\n?([\s\S]*?)\n?\s*```', r'\1', text).strip()


def extract_outermost_json_object(text: str) -> str:
    """
    Extract the outermost balanced JSON object using brace-depth counting.
    
    This avoids naive start/end matching that can span multiple disjoint objects.
    """
    start = text.find('{')
    if start == -1:
        raise ValueError(NO_JSON_FOUND_MESSAGE)
    
    depth = 0
    in_string = False
    escape = False
    
    for i in range(start, len(text)):
        char = text[i]
        
        if escape:
            escape = False
            continue
        
        if char == '\\' and in_string:
            escape = True
            continue
        
        if char == '"':
            in_string = not in_string
            continue
        
        if in_string:
            continue
        
        if char == '{':
            depth += 1
        elif char == '}':
            depth -= 1
            if depth == 0:
                return text[start:i + 1]
    
    raise ValueError(NO_JSON_FOUND_MESSAGE)

File: src/dotnet/Core/Infrastructure/Llm/Extraction/LlmOutputExtractor.cs
using System.Text.RegularExpressions;

namespace Rag.Candidates.Core.Infrastructure.Llm.Extraction;

/// <summary>
/// Pure static helper that cleans raw LLM output and extracts JSON content.
/// Handles markdown fences, thinking blocks, and preamble/postamble text.
/// </summary>
public static partial class LlmOutputExtractor
{
    private const string NoJsonFoundMessage = "No JSON object found in LLM output";

    /// <summary>
    /// Cleans raw LLM output and extracts the outermost JSON object as a string.
    /// </summary>
    public static string ExtractJson(string rawOutput)
    {
        if (string.IsNullOrWhiteSpace(rawOutput))
            throw new FormatException(NoJsonFoundMessage);

        var cleaned = StripThinkingBlocks(rawOutput);
        cleaned = StripMarkdownFences(cleaned);

        return ExtractOutermostJsonObject(cleaned);
    }

    /// <summary>
    /// Removes &lt;think&gt;...&lt;/think&gt; blocks (e.g. DeepSeek-R1).
    /// </summary>
    internal static string StripThinkingBlocks(string input)
    {
        return ThinkBlockRegex().Replace(input, string.Empty).Trim();
    }

    /// <summary>
    /// Removes markdown code fences: ```json ... ``` or ``` ... ```
    /// </summary>
    internal static string StripMarkdownFences(string input)
    {
        return MarkdownFenceRegex().Replace(input, static m => m.Groups[1].Value).Trim();
    }

    /// <summary>
    /// Extracts the outermost balanced JSON object from the string.
    /// Uses brace-depth counting rather than IndexOf/LastIndexOf to avoid
    /// spanning across multiple disjoint JSON objects.
    /// </summary>
    internal static string ExtractOutermostJsonObject(string input)
    {
        var start = input.IndexOf('{');
        if (start == -1)
            throw new FormatException(NoJsonFoundMessage);

        var depth = 0;
        var inString = false;
        var escape = false;

        for (var i = start; i < input.Length; i++)
        {
            var c = input[i];

            if (escape)
            {
                escape = false;
                continue;
            }

            if (c == '\\' && inString)
            {
                escape = true;
                continue;
            }

            if (c == '"')
            {
                inString = !inString;
                continue;
            }

            if (inString)
                continue;

            if (c == '{')
                depth++;
            else if (c == '}')
            {
                depth--;
                if (depth == 0)
                    return input[start..(i + 1)];
            }
        }

        throw new FormatException(NoJsonFoundMessage);
    }

    [GeneratedRegex(@"<think>[\s\S]*?</think>", RegexOptions.IgnoreCase | RegexOptions.Compiled)]
    private static partial Regex ThinkBlockRegex();

    [GeneratedRegex(@"```(?:json)?\s*\n?([\s\S]*?)\n?\s*```", RegexOptions.Compiled)]
    private static partial Regex MarkdownFenceRegex();
}
